Mathlib/RingTheory/PowerSeries/Basic.lean

```diff
@@ -2680,3 +2680,143 @@ theorem algebraMap_apply'' :
 #align power_series.algebra_map_apply'' PowerSeries.algebraMap_apply''
 
 end PowerSeries
+
+
+
+
+namespace PowerSeries
+
+open Nat hiding pow_succ pow_zero
+open Polynomial BigOperators Finset Finset.Nat
+
+/--
+`R⟦X⟧` is notation for `PowerSeries R`,
+the semiring of formal power series in one variable over a semiring `R`.
+-/
+scoped notation:9000 R "⟦X⟧" => PowerSeries R
+
+variable {R : Type*} [CommSemiring R]
+
+
+theorem trunc_trunc_of_le {n m} (f : R⟦X⟧) (hnm : n ≤ m := by rfl) :
+    trunc n ↑(trunc m f) = trunc n f := by
+  ext d
+  rw [coeff_trunc, coeff_trunc, Polynomial.coeff_coe]
+  split_ifs with h
+  · have : d < m := lt_of_lt_of_le h hnm
+    rw [coeff_trunc, if_pos this]
+  · rfl
+
+@[simp] theorem trunc_trunc {n} (f : R⟦X⟧) : trunc n ↑(trunc n f) = trunc n f :=
+  trunc_trunc_of_le f
+
+@[simp] theorem trunc_trunc_mul {n} (f g : R ⟦X⟧) :
+    trunc n ( (trunc n f) * g : R⟦X⟧ ) = trunc n ( f * g ) := by
+  ext m
+  rw [coeff_trunc, coeff_trunc]
+  split_ifs with h
+  · rw [coeff_mul, coeff_mul, sum_congr rfl]
+    intro _ hab
+    have ha := lt_of_le_of_lt (antidiagonal.fst_le hab) h
+    rw [Polynomial.coeff_coe, coeff_trunc, if_pos ha]
+  · rfl
+
+@[simp] theorem trunc_mul_trunc {n} (f g : R ⟦X⟧) :
+  trunc n ( f * (trunc n g) : R⟦X⟧ ) = trunc n ( f * g ) := by
+  rw [mul_comm, trunc_trunc_mul, mul_comm]
+
+theorem trunc_trunc_mul_trunc {n} (f g : R⟦X⟧) :
+    trunc n (trunc n f * trunc n g : R⟦X⟧) = trunc n (f * g) := by
+  rw [trunc_trunc_mul, trunc_mul_trunc]
+
+@[simp] theorem trunc_trunc_pow (f : R⟦X⟧) (n a : ℕ) :
+    trunc n ((trunc n f : R⟦X⟧) ^ a) = trunc n (f ^ a) := by
+  induction a with
+  | zero =>
+    rw [pow_zero, pow_zero]
+  | succ a ih =>
+    rw [pow_succ, pow_succ, trunc_trunc_mul, ←trunc_trunc_mul_trunc, ih, trunc_trunc_mul_trunc]
+
+theorem trunc_coe_eq_self {n} {f : R[X]} (hn : natDegree f < n) : trunc n (f : R⟦X⟧) = f := by
+  have this : support f ⊆ Ico 0 n
+  · calc
+      support f
+        ⊆ range (f.natDegree + 1)  := supp_subset_range_natDegree_succ
+      _ ⊆ range n                  := Iff.mpr range_subset hn
+      _ = Ico 0 n                  := by rw [range_eq_Ico]
+  nth_rw 2 [←sum_monomial_eq f]
+  rw [trunc, sum_eq_of_subset (hs := this), sum_congr rfl]
+  · intros
+    rw [Polynomial.coeff_coe]
+  · intros
+    apply monomial_zero_right
+
+theorem trunc_succ (f : R⟦X⟧) (n : ℕ) :
+    trunc n.succ f = trunc n f + Polynomial.monomial n (coeff R n f) := by
+  rw [trunc, Ico_zero_eq_range, sum_range_succ, trunc, Ico_zero_eq_range]
+
+
+/-- The function `coeff n : R⟦X⟧ → R` is continuous. I.e. `coeff n f` depends only on a sufficiently
+long truncation of the power series `f`.-/
+theorem coeff_stable {n m} {f : R⟦X⟧} (h : n.succ ≤ m := by rfl) :
```

Comment:
I'm not clear that `n.succ <= m := by rfl` is the best idea. Sure it makes your next proof a bit simpler, but in practice users will probably have hypotheses of the form `n < m` I would have thought.

Reply:
I agree. Thanks.